gr-iio/lib/device_source_impl.cc

```diff
@@ -311,7 +318,8 @@ int device_source_impl::work(int noutput_items,
     unsigned long items = std::min(items_in_buffer, (unsigned long)noutput_items);
 
     for (size_t i = 0; i < output_items.size(); i++)
-        channel_read(channel_list[i], output_items[i], items * sizeof(short));
+        channel_read(
+            channel_list[i], output_items[i], items * channel_size(channel_list[i]));
```

Comment:
Since `channel_size()` calls `iio_channel_get_data_format()` in `work()`, can we assume that the iio call is very low cost?

Reply:
Yes, it is a simple getter from memory - https://github.com/analogdevicesinc/libiio/blob/main/channel.c#L442-L446 
Alternatively one could get channel size directly from the signature with int gr::io_signature::sizeof_stream_item(int index)

Reply:
It's probably fine as is. A couple of pointer indirections per block of samples shouldn't make a difference, and you'd have at least one indirection with a cached vector of sizes.

Reply:
@willcode jep, that's why I hoped for `constexpr` (which was not very clever of me!); anyways, the overhead here is that `get_channel_data_format` is a dynamically loaded&linked symbol, so we're actually doing the full call overhead here, which probably is a fair bit more expensive than the pointer derefence involved. It's sensible that libiio has the channel struct as opaque data structure (to avoid people fumbling about in IIO-internal data structures), but unless you can do `constexpr` or `__attribute__((pure))` or `__attribute__((const))` on the actual object. 

(The latter two are supported when built with GCC, so that's why the constexpr worked. I'm *kind of worried* that they actually shouldn't be there, now having slept about that, but way we currently deal with things should be safe, kind of, and it would be an IIO bug, not a gr-iio bug if it weren't: 

https://github.com/analogdevicesinc/libiio/blob/main/include/iio/iio.h#L1498-L1499C34

```C
__api __check_ret __cnst const struct iio_data_format * iio_channel_get_data_format(
		const struct iio_channel *chn);
```

`__cnst` expands to `__attribute__((const))`, but that tells GCC "If you pass in the same pointer (not the identical object pointed to!) twice, you get the same result"; but that's not actually the case, because a pointer to the same address might point to a different object next invocation, so the data format might be different. See [GCC's warning about use with pointers, as well](https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#index-const-function-attribute); will file a bug against libiio)

Reply:
I'm going to interpret that as "ok, we'll see how it works".